drivers/xen/gnttab.c

```diff
@@ -324,46 +331,150 @@ const char *gnttabop_error(int16_t status)
 	}
 }
 
-static int gnttab_init(void)
+static int map_grant_frames(unsigned int start_frame, unsigned int end_frame)
 {
-	grant_ref_t gref;
-	struct xen_add_to_physmap xatp;
+	int rc, i;
+	unsigned int nr_frames = end_frame - start_frame;
 	struct gnttab_setup_table setup;
-	xen_pfn_t frames[NR_GRANT_FRAMES];
-	int rc = 0, i;
-
-	/* Will be taken/given during gnt_refs allocation/release */
-	k_sem_init(&gnttab.sem, 0, NR_GRANT_ENTRIES - GNTTAB_NR_RESERVED_ENTRIES);
-
-	for (
-		gref = GNTTAB_NR_RESERVED_ENTRIES;
-		gref < NR_GRANT_ENTRIES;
-		gref++
-	    ) {
-		put_free_entry(gref);
+	xen_pfn_t *frames;
+
+	if (gnttab.nr_grant_frames + nr_frames > gnttab.max_grant_frames) {
+		LOG_ERR("Reached max number of Xen grant frames");
+		return -ENOMEM;
 	}
 
-	for (i = 0; i < NR_GRANT_FRAMES; i++) {
+	for (i = start_frame; i < end_frame; i++) {
+		struct xen_add_to_physmap xatp;
+
+		/* Stage 2 mapping for first frame */
 		xatp.domid = DOMID_SELF;
 		xatp.size = 0;
 		xatp.space = XENMAPSPACE_grant_table;
 		xatp.idx = i;
 		xatp.gpfn = xen_virt_to_gfn(Z_TOPLEVEL_ROM_NAME(grant_tables).phys_addr) + i;
 		rc = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
-		__ASSERT(!rc, "add_to_physmap failed; status = %d\n", rc);
+		if (rc) {
+			LOG_ERR("add_to_physmap failed; status = %d\n", rc);
+			return rc;
+		}
+	}
+
+	gnttab.nr_grant_frames += nr_frames;
+
+	frames = k_calloc(gnttab.nr_grant_frames, sizeof(*frames));
+	if (!frames) {
+		LOG_ERR("Failed to allocate memory for frames");
+		return -ENOMEM;
 	}
 
+	/* Frames are mapped, so we can setup grant table */
```

Comment:
Oh, this function not only maps frame but also sets it up. In this case it should have this reflected in its name

Reply:
Divided into 2 separate functions.